crates/acvm/src/pwg/binary.rs

```diff
@@ -0,0 +1,236 @@
+                        if (arith.mul_terms[0].0 + arith.linear_combinations[0].0).is_zero() {
+                            x = Some(0);
+                        }
+                    } else if (arith.mul_terms[0].0 + arith.q_c).is_zero()
+                        && arith.mul_terms[0].0 == arith.linear_combinations[0].0
+                    {
+                        x = Some(0);
+                    }
+                }
+            }
+        } else if arith.mul_terms.is_empty() && arith.linear_combinations.len() == 2 {
+            //x=y
+            let z = if self.is_boolean(&arith.linear_combinations[0].1)
+                && !self.is_boolean(&arith.linear_combinations[1].1)
+            {
+                Some(1)
+            } else if self.is_boolean(&arith.linear_combinations[1].1)
+                && !self.is_boolean(&arith.linear_combinations[0].1)
+            {
+                Some(0)
+            } else {
+                None
+            };
+            if z.is_some() {
+                if arith.q_c.is_zero() {
+                    if (arith.linear_combinations[0].0 + arith.linear_combinations[1].0).is_zero() {
+                        x = z;
+                    }
+                } else if (arith.q_c + arith.linear_combinations[1].0).is_zero()
+                    && arith.linear_combinations[0].0 == arith.linear_combinations[1].0
+                {
+                    x = Some(0);
+                }
+            }
+        } else if arith.mul_terms.is_empty() && arith.linear_combinations.len() > 2 {
+            //"binary" gates 'optimised' by the optimiser should have an intermediate variable and a bunch of booleans
+            let mut max = BigUint::from_bytes_be(&arith.q_c.to_bytes());
+            for (i, lin) in arith.linear_combinations.iter().enumerate() {
+                if let Some(v) = self.get_max_value(&lin.1) {
+                    max += v * BigUint::from_bytes_be(&lin.0.to_bytes());
+                } else if x.is_some() {
+                    x = None;
+                    break;
+                } else {
+                    x = Some(i);
+                }
+            }
+            if max < FieldElement::modulus()
+                && x.is_some()
+                && arith.linear_combinations[x.unwrap()].0 == -FieldElement::one()
+            {
+                self.positive_witness.insert(arith.linear_combinations[x.unwrap()].1, max);
+                x = None;
+            }
+        }
+        if let Some(x) = x {
+            self.binary_witness.insert(arith.linear_combinations[x].1);
+        }
+    }
+
+    // identify boolean and inverse constraints in the gate
+    pub fn identify_binaries(&mut self, gate: &Gate) {
+        match gate {
+            Gate::Directive(Directive::Invert { x, result }) => {
+                self.invert_witness.insert(*x, *result);
+            }
+            Gate::Arithmetic(a) => {
+                self.identify_booleans(a);
```

Comment:
When is this branch triggered?

seems identify_binaries is called when it is not an arithmetic gate

Reply:
Indeed, but still the function is correct like this, using unreachable! would make it fit for the current use and less versatile.